app/controllers/concerns/authentication.rb

```diff
@@ -96,10 +110,53 @@ def query_token_authenticator(query_token)
     end
   end
 
+  def http_cookie_authenticator(session_id)
+    session = current_account.sessions.for_environment(current_environment, strict: current_environment.nil?)
+                                      .preload(:token, :bearer)
+                                      .find_by(
+                                        id: session_id,
+                                      )
+
+    @current_http_scheme = :session
+    @current_http_token  = nil
+
+    raise Keygen::Error::UnauthorizedError.new(code: 'SESSION_INVALID') if
+      session.nil? || session.token.nil? || session.bearer.nil?
+
+    raise Keygen::Error::UnauthorizedError.new(code: 'SESSION_EXPIRED', detail: 'Session is expired') if
+      session.expired?
+
+    raise Keygen::Error::ForbiddenError.new(code: 'USER_BANNED', detail: 'User is banned') if
+      session.bearer.respond_to?(:banned?) && session.bearer.banned?
+
+    case
+    when session.bearer.has_role?(:license)
+      raise Keygen::Error::ForbiddenError.new(code: 'SESSION_NOT_ALLOWED', detail: 'Session authentication is not allowed by policy') unless
+        session.bearer.supports_session_auth?
+    end
+
+    if session.last_used_at.nil? || session.last_used_at.before?(1.hour.ago)
+      session.update(
+        expiry: session.expiry + 1.hour, # extend expiry while in use until MAX_AGE
```

Comment:
This should only extend expiry when it expires within a certain duration e.g. 1 hour.